### PR TITLE
Add `@save`

### PR DESCRIPTION
**Pull request by LilithHafner**
Implement `@save` debugging tool similar to `@show`, but instead of printing, saves the captured values to globals in Main.

````
    @save variable
    @save (variable | variable=expression)...

Save `variable` for manual user inspection.

When using the default [`ConsoleLogger`](@ref), Sets `Main.varaible` equal to the result
of `expression`, or the value of `variable` if `expression` is not provided.

Behavior may be overridden by a custom logging backend. See [`@logmsg`](@ref) for more info.

# Examples

```jldoctest
julia> function f(x)
           x -= 1
           @save x
           1/x
       end
f (generic function with 1 method)

julia> f(1)
Inf

julia> x
0
```
````

---

Passes through Logging so that results can be aggregated or redirected by custom logging backends. e.g. you might want to save all the results of an `@save` callsite instead of only the most recent.

### CODE REVIEW

**Comment by fredrikekre** [comment]
This looks like https://juliadebug.github.io/Infiltrator.jl/stable/API/#Infiltrator.@exfiltrate. I don't think using the log infrastructure for this is a good idea.

File: test/corelogging.jl [hunk 1]
```diff
@@ -467,3 +467,43 @@ end
 for (k, v) in pairs(original_env)
     ENV[k] = v
 end
+
+# @save
+module AtSaveTestModule
+    using Test
+    @test !isdefined(Main, :xc44f6e16f91ba48824afec66744d3faf_1)
```
Comment (KristofferC) [inline test/corelogging.jl:474]: These are excessively long, makes it hard to read.

**Comment by KristofferC** [comment]
As been said, Infiltrator does this (but better). The name is very generic and should probably not be exported. Should probably use the active REPL module.

Doesn't feel worth it to take this pretty exclusive symbol name for a worse version of a single feature from an existing package.

**Comment by c42f** [comment]
I think something like this is a good idea.

Using the logging system is not equivalent to `@exfiltrate` in several important ways.

One of the key things you get when using logging is a full trace of the behavior of the system over time. Potentially from several log sites. This is a lot more information than saving a single variable, and can often be what you need to understand a complex system. Eg, there's a reason distributed systems people have such sophisticated logging systems. But it applies in simpler cases.

If I was to write a blog post on this, it might be called "The unreasonable effectiveness of printf debugging" or something. Clickbait? Maybe, but also true.

> Passes through Logging so that results can be aggregated or redirected by custom logging backends. e.g. you might want to save all the results of an @save callsite instead of only the most recent.

To expand on this, one might want to plot the values of the variable over time. Maybe the value from one `@save` callsite or several different ones. I've long wanted `@show` to work like this and regret I didn't push through with doing it in julia-0.7. (There's an off chance that it still might be possible to make `@show` do this without breaking anything important. But I bet someone, somewhere, depends on it working just as it does now.)

In terms of package dependencies, `Infiltrator` is often a bit annoying because it introduces a dependency. There's hacky ways around this, but having a logging frontend like `@save` in Base would be extremely convenient.

---

I do think there's some work to do here
- The logging system `message` field has the wrong semantics to communicate a variable. The right place to put arbitrary variables is in the key-value pairs. (The message is probably just something empty? idk?)
- All arguments to `@save` should end up in a single logging message, as they represent program values at a single point in time.
- Ideally we'd prove out the user experience of the aggregation/plotting use cases before we commit to a name for this.

I know that last one is a big ask. But I think having a logging backend which does at least some of that (outside Base, but perhaps loaded by default for interactive sessions) will eventually make a version of this really compelling.

**Comment by vchuravy** [comment]
So I think for me the biggest oddity is that this works by setting globals in Main.

My gut-check here would be to use structured logging to exfiltrate data (as you currently can)
and then provide better ways for accessing them by post-processing.

So I would totally love to have something that collects log messages in the REPL and I can go back in time and access the data in the structure log message. But evaling globals into main strikes me as the wrong approach, especially since there might be arbitrary conflicts between `@save` messages in packages.

**Comment by StefanKarpinski** [comment]
What I've found to be super useful in the past for this kind of debugging is to emit csv lines with these fields: timestamp, node id, process id, variable name, value. Then you load one or more such file into a database and run queries on it. Having easy support for that style of debugging would be very useful but I'm not sure it belongs in Base. Using the logging infrastructure seems nice but parsing logs seems annoying unless the proposal is a logger that matches what I described.

**Comment by c42f** [comment]
> Having easy support for that style of debugging would be very useful

Exactly the use case I'm imagining! It's very powerful.

> but I'm not sure it belongs in Base.

Yep, only the logging *frontend* to support it should be in Base (maybe).  That's in line with the other logging infrastructure like `@info` - basically similar and has similar advantages in that packages don't pull in a dependency to do what amounts to "printf debugging"

**Comment by fredrikekre** [comment]
> What I've found to be super useful in the past for this kind of debugging is to emit csv lines with these fields: timestamp, node id, process id, variable name, value.

These thing is pretty easy to achieve already with LoggingExtras + e.g. the [JSON logging format](https://github.com/JuliaLogging/LoggingFormats.jl?tab=readme-ov-file#json-output-log-events-as-json).

The `handle_message` changes in this PR really seems like the kind of specialization you do for a new logger type and dispatch on it, `SaveLogger` or something. I don't really understand why the existing macros can't be used here.

**Comment by c42f** [comment]
> I don't really understand why the existing macros can't be used here.

Really, the only reason is ergonomics.

Unfortunately `@info` has a `message` argument with the wrong semantics to put values into.  `@info x` doesn't really make sense - it should be `@info "" x`.  This empty message kind of pains me every time I type it. And I type it a lot.

Also I feel the default logger installed in the REPL should have "some sensible behavior" - when in interactive mode - for allowing users to slice and dice log messages.

What I'd really like is to type `@show x` in my code (or `@save x`, or whatever ... a better verb?), and have some simple way to get at that value which is part of the default REPL experience. We have nearly all the infrastructure we need.

Perhaps another good way to attack this is to forget the name of the frontend for now - just use `@info "" x` for now - and focus on proving that we can have a really good backend for this (can be in a package such as TerminalLoggers as a PoC).

I feel the value of a short name for this would be more obvious once we have good tooling.

**Comment by tecosaur** [comment]
Personally I feel like this has more value in the context of things like workspaces, and I'm not sure I see much benefit with this over `Core.eval(Main, :(x = $x))`. I've only found this useful a few times, and so I'm wary of assigning the name `@save` to this.

**Pull request closed without merging**